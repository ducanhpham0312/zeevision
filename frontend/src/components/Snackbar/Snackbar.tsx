import * as React from "react";
import { Transition } from "react-transition-group";
import { twMerge } from "tailwind-merge";
import CheckRoundedIcon from "@mui/icons-material/CheckRounded";
import ErrorIcon from "@mui/icons-material/Error";
import CloseIcon from "@mui/icons-material/Close";
import { Snackbar as SnackbarMUI } from "@mui/base/Snackbar";
import { SnackMessageType, useUIStore } from "../../contexts/useUIStore";
import { SnackbarCloseReason } from "@mui/base";
import { VariantProps, tv } from "tailwind-variants";

const SNACKBAR_AUTO_CLOSE_MS = 5000;

type TransitionStatus =
  | "entering"
  | "entered"
  | "exiting"
  | "exited"
  | "unmounted";

interface SnackbarContentProps
  extends VariantProps<typeof snackbarContentVariant> {
  /**
   * The message to be displayed within the snackbar.
   */
  message: SnackMessageType["message"];
  /**
   * The title of the snackbar.
   */
  title: SnackMessageType["title"];
  /**
   * The type of the snackbar which determines its appearance and icon.
   * For example, 'error' or 'success'.
   */
  type: SnackMessageType["type"];
  /**
   * The status of the transition states which affect snackbar visibility and positioning.
   */
  status: TransitionStatus;
  /**
   * A mutable ref object which is used to manage the component's
   * render cycle and avoid potential issues during transition phases.
   */
  nodeRef?: React.MutableRefObject<null>;

  messageNode?: React.ReactNode;

  className?: string;
  /**
   * A function that handles the closing of the snackbar.
   * It can be triggered by a user action (e.g., clicking a button)
   * or programmatically, depending on the implementation.
   *
   * @param _: Event or null - The triggering event or null if closed programmatically.
   * @param reason: SnackbarCloseReason - Describes why the snackbar is being closed.
   */
  handleClose: (_: Event | null, reason: SnackbarCloseReason) => void;
}

export function Snackbar() {
  const { snackbarContent, closeSnackBar } = useUIStore();
  const { messageNode, open, message, title, type } = snackbarContent;
  const [exited, setExited] = React.useState(true);
  const timeoutId = React.useRef<number | undefined>();
  const nodeRef = React.useRef(null);

  const handleClose = (_: Event | null, reason: SnackbarCloseReason) => {
    if (reason === "clickaway") {
      return;
    }
    closeSnackBar();
  };

  const handleOnEnter = () => {
    setExited(false);
  };

  const handleOnExited = () => {
    setExited(true);
  };

  React.useEffect(() => {
    if (open) {
      if (timeoutId.current) {
        clearTimeout(timeoutId.current);
      }
      timeoutId.current = setTimeout(() => {
        closeSnackBar();
      }, SNACKBAR_AUTO_CLOSE_MS) as unknown as number;
    }

    return () => {
      if (timeoutId.current) {
        clearTimeout(timeoutId.current);
      }
    };
  }, [snackbarContent, closeSnackBar, open]);

  return (
    <SnackbarMUI
      className="fixed bottom-4 right-4 z-50 flex min-w-[360px] max-w-[560px]"
      open={open}
      onClose={handleClose as () => void}
      exited={exited}
    >
      <Transition
        timeout={{ enter: 400, exit: 400 }}
        in={open}
        appear
        unmountOnExit
        onEnter={handleOnEnter}
        onExited={handleOnExited}
        nodeRef={nodeRef}
      >
        {(status) => (
          <SnackbarContent
            type={type}
            title={title}
            message={message}
            messageNode={messageNode}
            status={status}
            handleClose={handleClose}
            nodeRef={nodeRef}
          />
        )}
      </Transition>
    </SnackbarMUI>
  );
}

export function SnackbarContent({
  type,
  status,
  title,
  message,
  handleClose,
  nodeRef,
  className,
  messageNode,
}: SnackbarContentProps) {
  return (
    <div
      className={twMerge(snackbarContentVariant({ type }), className)}
      style={{
        transform: positioningStyles[status],
        transition: "transform 300ms ease",
        alignContent: "start",
      }}
      ref={nodeRef}
    >
      {type === "success" ? (
        <CheckRoundedIcon
          className="success-icon"
          sx={{
            color: "white",
            flexShrink: 0,
            width: "1.25rem",
            height: "1.5rem",
          }}
        />
      ) : (
        <ErrorIcon
          className="error-icon"
          sx={{
            color: "white",
            flexShrink: 0,
            width: "1.25rem",
            height: "1.5rem",
          }}
        />
      )}
      <div className="flex flex-col max-w-full">
        <p className="snackbar-title">{title}</p>
        {messageNode ? (
          messageNode
        ) : (
          <p className="snackbar-description">{message}</p>
        )}
      </div>

      <button
        className="close-button"
        style={{ all: "unset", height: "24px" }}
        onClick={handleClose as () => void}
      >
        <CloseIcon sx={{ fontSize: "20px" }} className="snackbar-close-icon" />
      </button>
    </div>
  );
}

const snackbarContentVariant = tv({
  base: "flex gap-2 p-4 overflow-hidden rounded text-white",
  variants: {
    type: {
      success: "bg-sucess",
      error: "bg-error",
    },
  },
});

const positioningStyles = {
  entering: "translateX(0)",
  entered: "translateX(0)",
  exiting: "translateX(110%)",
  exited: "translateX(110%)",
  unmounted: "translateX(110%)",
};
