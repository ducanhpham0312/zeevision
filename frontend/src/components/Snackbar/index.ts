export { Snackbar, SnackbarContent } from "./Snackbar";
