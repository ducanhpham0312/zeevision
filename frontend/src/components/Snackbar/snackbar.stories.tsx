import type { Meta, StoryObj } from "@storybook/react";
import { SnackbarContent } from ".";

const meta = {
  title: "Assets/Snackbar",
  component: SnackbarContent,
  parameters: {
    layout: "centered",
  },
  tags: ["autodocs"],
  argTypes: {},
} satisfies Meta<typeof SnackbarContent>;

export default meta;
type Story = StoryObj<typeof meta>;

export const Success: Story = {
  args: {
    status: "entering",
    title: "Stoybook success snackbar",
    message: "Everything was sent to the desired address.",
    type: "success",
    handleClose: () => console.log("Snackbar handle close called"),
  },
};

export const Error: Story = {
  args: {
    status: "entering",
    title: "Stoybook error snackbar",
    message: "Everything was not sent to the desired address.",
    type: "error",
    handleClose: () => console.log("Snackbar handle close called"),
  },
};
