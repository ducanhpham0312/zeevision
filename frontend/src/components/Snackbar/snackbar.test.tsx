import { act, render } from "@testing-library/react";
import "@testing-library/jest-dom";
import { Snackbar } from "./Snackbar";

const mockCloseSnackFunction = jest.fn();

// Mocking the context hook
jest.mock("../../contexts/useUIStore", () => ({
  useUIStore: () => ({
    snackMessage: {
      open: true,
      message: "Test message",
      title: "Test title",
      type: "success",
    },
    closeSnackMessage: mockCloseSnackFunction,
  }),
}));

jest.useFakeTimers();

describe("Snackbar Component", () => {
  test("renders and displays the message and title", () => {
    const { getByText } = render(<Snackbar />);
    expect(getByText("Test message")).toBeInTheDocument();
    expect(getByText("Test title")).toBeInTheDocument();
  });

  it("calls handleClose when the close button is clicked", async () => {
    render(<Snackbar />);
    // await userEvent.click(screen)
    expect(true).toBe(true);
  });

  test("should close the Snackbar after timeout", () => {

    render(<Snackbar />);
    act(() => {
      jest.advanceTimersByTime(3000);
    });

    expect(mockCloseSnackFunction).toHaveBeenCalled();
  });
});
