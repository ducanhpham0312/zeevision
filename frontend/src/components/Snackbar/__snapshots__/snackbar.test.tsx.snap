// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snackbar Component render correctly 1`] = `
<DocumentFragment>
  <div
    class="MuiSnackbar-root fixed bottom-4 right-4 z-50 flex min-w-[360px] max-w-[560px]"
    role="presentation"
  >
    <div
      class="flex gap-2 p-4 overflow-hidden rounded text-white bg-sucess"
      style="transform: translateX(0); transition: transform 300ms ease; align-content: start;"
    >
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium success-icon css-c1qtpg-MuiSvgIcon-root"
        data-testid="CheckRoundedIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M9 16.17 5.53 12.7a.9959.9959 0 0 0-1.41 0c-.39.39-.39 1.02 0 1.41l4.18 4.18c.39.39 1.02.39 1.41 0L20.29 7.71c.39-.39.39-1.02 0-1.41a.9959.9959 0 0 0-1.41 0L9 16.17z"
        />
      </svg>
      <div
        class="flex max-w-full"
      >
        <p
          class="snackbar-title"
        >
          Default title
        </p>
        <p
          class="snackbar-description"
        >
          Default message
        </p>
      </div>
      <button
        class="close-button"
        style="all: unset; height: 24px;"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium snackbar-close-icon css-1lrr7yx-MuiSvgIcon-root"
          data-testid="CloseIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
          />
        </svg>
      </button>
    </div>
  </div>
</DocumentFragment>
`;
